### PR TITLE
java: async-profiler v2.5 and CPU mode with fdtransfer

### PR DESCRIPTION
**Pull request by Jongy**
## Description
Bases on https://github.com/jvm-profiling-tools/async-profiler/pull/411 to use AP's CPU mode.

## Motivation and Context
We'll have kernel stacks for profiled Java processes :)

## How Has This Been Tested?
Locally. ~~I'll add some test as well.~~ added a test

## Checklist:
- [ ] I have read the **CONTRIBUTING** document.
- [x] Wait for the next AP release (haven't updated it yet)
- [ ] I have updated the relevant documentation.
- [x] I have added tests for new logic.

### CODE REVIEW

**Comment by Jongy** [comment]
Default to `itimer` so we can use the CPU mode manually in some deployments, until we gain confidence in it.

**Comment by michelhe** [comment]
explain why?

File: gprofiler/utils.py [hunk 1]
```diff
@@ -164,13 +164,23 @@ def run_process(
...
-                stdout, stderr = process.communicate(timeout=timeout)
+                assert timeout is None
+                if communicate:
+                    # wait for stderr & stdout to be closed
+                    stdout, stderr = process.communicate(timeout=timeout)
+                else:
+                    # just wait for the process to exit
+                    process.wait()
             else:
+                assert communicate
```
Comment (michelhe) [inline gprofiler/utils.py:183]: assert with message?

**Comment by michelhe** [comment]
AP build seems to fail in CI

**Review by michelhe — CHANGES_REQUESTED** [review]
.

**Comment by Jongy** [comment]
```
$ git range-diff v2.5..v2.5g1 v2.0..v2.0g5
1:  5f90814 ! 1:  ab1b8c2 Replace semicolon in method descriptors to vertical bar, since a semicolon is already used to separate between methods
    @@ src/profiler.cpp: void Profiler::dumpCollapsed(std::ostream& out, Arguments& arg
     +            std::replace(frame_name.begin(), frame_name.end(), ';', '|');
                  out << frame_name << (j == 0 ? ' ' : ';');
              }
    -         out << samples << "\n";
    +         out << (args._counter == COUNTER_SAMPLES ? (*it)->samples : (*it)->counter) << "\n";
2:  b3ff118 ! 2:  9692c57 Remove the "Profiling started" message
    @@ src/profiler.cpp: Error Profiler::runInternal(Arguments& args, std::ostream& out
                  if (error) {
                      return error;
                  }
    --            out << "Profiling started\n";
    +-            out << "Profiling started" << std::endl;
                  break;
              }
              case ACTION_STOP: {
3:  d10c12a < -:  ------- Remove the "Profiling stopped" message
-:  ------- > 3:  51447a8 Remove the "Profiling stopped" message
4:  d7f1029 ! 4:  6483566 Compile statically with libstdc++
    @@ Makefile: build:
     -	$(CXX) $(CXXFLAGS) -DPROFILER_VERSION=\"$(PROFILER_VERSION)\" $(INCLUDES) -fPIC -shared -o $@ $(SOURCES) $(LIBS)
     +	$(CXX) $(CXXFLAGS) -DPROFILER_VERSION=\"$(PROFILER_VERSION)\" $(INCLUDES) -fPIC -shared -static-libstdc++ -o $@ $(SOURCES) $(LIBS)
      
    - build/$(JATTACH): src/jattach/*.c src/jattach/*.h
    - 	$(CC) $(CFLAGS) -DJATTACH_VERSION=\"$(PROFILER_VERSION)-ap\" -o $@ src/jattach/*.c
    + build/$(JATTACH): src/jattach/jattach.c
    + 	$(CC) $(CFLAGS) -DJATTACH_VERSION=\"$(PROFILER_VERSION)-ap\" -o $@ $^
5:  b9645a7 < -:  ------- Build jattach statically
-:  ------- > 5:  9d29169 Build jattach statically
6:  e70054f ! 6:  3f1505b Enocde buildid+offset of unknown symbols in the frame (#1)
    @@ src/arguments.h: class Arguments {
              _buf(NULL),
     @@ src/arguments.h: class Arguments {
              _end(NULL),
    -         _title(NULL),
    +         _title("Flame Graph"),
              _minwidth(0),
     -        _reverse(false) {
     +        _reverse(false),
    @@ src/codeCache.cpp
      #include "codeCache.h"
      
      
    -@@ src/codeCache.cpp: NativeCodeCache::NativeCodeCache(const char* name, short lib_index, const void*
    -     _lib_index = lib_index;
    +@@ src/codeCache.cpp: NativeCodeCache::NativeCodeCache(const char* name, const void* min_address, cons
    +     _name = strdup(name);
          _min_address = min_address;
          _max_address = max_address;
     +    _build_id = NULL;
    @@ src/codeCache.cpp: NativeCodeCache::NativeCodeCache(const char* name, short lib_
      NativeCodeCache::~NativeCodeCache() {
     +    free(_build_id);
          for (int i = 0; i < _count; i++) {
    -         free((char*)_blobs[i]._method - LIB_INDEX_OFFSET);
    +         free(_blobs[i]._method);
          }
     @@ src/codeCache.cpp: void NativeCodeCache::sort() {
          if (_max_address == NO_MAX_ADDRESS) _max_address = _blobs[_count - 1]._end;
    @@ src/codeCache.h
      
      #define NO_MIN_ADDRESS  ((const void*)-1)
      #define NO_MAX_ADDRESS  ((const void*)0)
    -@@ src/codeCache.h: class NativeCodeCache : public CodeCache {
    +@@ src/codeCache.h: class CodeCache {
    + class NativeCodeCache : public CodeCache {
        private:
          char* _name;
    -     short _lib_index;
     +    char *_build_id;
      
    -     static char* encodeLibrarySymbol(const char* name, short lib_index);
    - 
    +   public:
    +     NativeCodeCache(const char* name,
     @@ src/codeCache.h: class NativeCodeCache : public CodeCache {
      
          void add(const void* start, int length, const char* name, bool update_bounds = false);
    @@ src/profiler.cpp: Error Profiler::start(Arguments& args, bool reset) {
     
      ## src/profiler.h ##
     @@ src/profiler.h: class Profiler {
    +     CStack _cstack;
          bool _add_thread_frame;
    -     bool _add_sched_frame;
          bool _update_thread_names;
     +    bool _add_build_ids;
          volatile bool _thread_events_state;
7:  02cb785 < -:  ------- Build fdtransfer statically

```

**Comment by Jongy** [comment]
Re-ran, now it'll build successfully.

**Review by michelhe — APPROVED** [review]

File: gprofiler/utils.py (hunk 1 above)
Comment (Jongy) [reply gprofiler/utils.py]: Done

File: gprofiler/profilers/perf.py [hunk 2]
```diff
@@ -27,6 +27,8 @@ def perf_path() -> str:
 class PerfProcess:
     _dump_timeout_s = 5
     _poll_timeout_s = 5
+    _mmap_size_fp = 129  # default number of pages used by "perf record" when perf_event_mlock_kb=516
+    _mmap_size_dwarf = 257  # doubling the previous
```
Comment (michelhe) [inline gprofiler/profilers/perf.py:31]: cleaner
```
_mmap_sizes = { 'fp': 129, 'dwarf': 257 }

....

str(self._mmap_sizes[self._type])
```

**Review by michelhe — APPROVED** [review]
kool

**Pull request merged**